src/session.rs

```diff
@@ -263,31 +270,48 @@ impl Session {
                 )
             })?;
 
-            inner.rc(handshake(inner.raw, stream))
+            inner.rc(handshake(inner.raw, stream.as_ref()))
         }
     }
 
-    /// The session takes ownership of the socket provided.
-    /// You may use the tcp_stream() method to obtain a reference
-    /// to it later.
+    /// The session takes ownership of the stream provided.
+    /// You may use the tcp_stream() method to obtain the raw fd later.
+    ///
+    /// It is also highly recommended that the stream provided is not used
+    /// concurrently elsewhere for the duration of this session as it may
+    /// interfere with the protocol.
+    #[cfg(unix)]
+    pub fn set_tcp_stream<S: 'static + AsRawFd>(&mut self, stream: S) {
```

Comment:
not a real concern, just an observation: I'm slightly surprised that the lifetime is needed here

Reply:
I think this is related to: https://stackoverflow.com/questions/29740488/parameter-type-may-not-live-long-enough .

So the bound doesn't allow that the trait object stored inside the session outlives everything it references to. I don't know if there is a better way, to me `'static` seems ok.